sim/battle.js

```diff
@@ -1548,6 +1548,19 @@ class Battle extends Dex.ModdedDex {
 					} else {
 						pokemon.lastAttackedBy = null;
 					}
+					if (this.gen >= 7) pokemon.knownType = true; // If it was an illusion, it's not any more
+				}
+
+				if (this.gen >= 7) {
+					// In Gen 7, the real type of every Pokemon is visible to all players via the bottom screen while making choices
+					let realTypeString = (pokemon.illusion || pokemon).getTypes(true).join('/');
+					let templateTypeString = (pokemon.illusion || pokemon).baseTemplate.types.join('/');
```

Comment:
After more thought, I think I'd be okay with this only in situations where the shown type otherwise doesn't match the user-visible type, and was changed. So instead of this being `baseTemplate`, it would actually keep track of what the client thinks the type is, and only emit the type update if there's a mismatch.

Reply:
Bleh, in that case this actually would need to be combined with #4654 since that changes how `'-formechange'` messages are sent.

Reply:
Should I rebase this branch on that one?

Reply:
Rather, should I combine them now and close that other pull request, or should I wait for the other one to be merged first, _then_ rebase?

Reply:
Never mind, I figured out a good solution that doesn't rely on any changes made in the other PR.